src/sema/function_annotation.rs

```diff
@@ -0,0 +1,184 @@
+                        &note.value,
+                        context,
+                        ns,
+                        symtable,
+                        &mut diagnostics,
+                        ResolveTo::Type(&ty),
+                    ) {
+                        if let Ok(expr) = expr.cast(&expr.loc(), &ty, true, ns, &mut diagnostics) {
+                            if let Some(prev) = &space {
+                                diagnostics.push(Diagnostic::error_with_note(
+                                    loc,
+                                    "duplicate @space annotation for constructor".into(),
+                                    *prev,
+                                    "previous @space".into(),
+                                ));
+                            } else {
+                                space = Some(loc);
+                                used_variable(ns, &expr, symtable);
+                                parsed_annotations.push(ConstructorAnnotation::Space(expr));
+                            }
+                        }
+                    }
+                    continue;
+                }
+                "payer" => {
+                    let ty = Type::Address(false);
+                    let loc = note.loc;
+
+                    if let Ok(expr) = expression(
+                        &note.value,
+                        context,
+                        ns,
+                        symtable,
+                        &mut diagnostics,
+                        ResolveTo::Type(&ty),
+                    ) {
+                        if let Ok(expr) = expr.cast(&expr.loc(), &ty, true, ns, &mut diagnostics) {
+                            if let Some(prev) = &payer {
+                                diagnostics.push(Diagnostic::error_with_note(
+                                    loc,
+                                    "duplicate @payer annotation for constructor".into(),
+                                    *prev,
+                                    "previous @payer".into(),
+                                ));
+                            } else {
+                                payer = Some(loc);
+                                used_variable(ns, &expr, symtable);
+                                parsed_annotations.push(ConstructorAnnotation::Payer(expr));
+                            }
+                        }
+                    }
+                    continue;
+                }
+                _ => (),
+            }
+        }
+
+        diagnostics.push(Diagnostic::error(
+            note.loc,
+            format!(
+                "unknown annotation {} for {}",
+                note.id.name, ns.functions[function_no].ty
+            ),
+        ));
+    }
+
+    if !parsed_annotations.is_empty()
+        && diagnostics.is_empty()
+        && !parsed_annotations
+            .iter()
+            .any(|annotation| matches!(annotation, ConstructorAnnotation::Payer(..)))
```

Comment:
You don't need to search for the payer in the vector. There can just be a boolean that you can set to true when you see the payer tag.

Reply:
We can reuse `payer.is_none()` for this